components/vr/webvr_thread.rs

```diff
@@ -0,0 +1,332 @@
+            },
+            Err(msg) => {
+                sender.send(Err(msg.into())).unwrap();
+            }
+        }
+    }
+
+    fn handle_exit_present(&mut self,
+                         pipeline: PipelineId,
+                         device_id: u64,
+                         sender: Option<IpcSender<WebVRResult<()>>>) {
+        match self.access_check(pipeline, device_id).map(|d| d.clone()) {
+            Ok(device) => {
+                self.presenting.remove(&device_id);
+                if let Some(sender) = sender {
+                    sender.send(Ok(())).unwrap();
+                }
+                let data = device.borrow().display_data();
+                self.notify_event(VRDisplayEvent::PresentChange(data, false));
+            },
+            Err(msg) => {
+                if let Some(sender) = sender {
+                    sender.send(Err(msg.into())).unwrap();
+                }
+            }
+        }
+    }
+
+    fn handle_create_compositor(&mut self, device_id: u64) {
+        let compositor = self.service.get_device(device_id).map(|d| WebVRCompositor(d.as_ptr()));
+        self.vr_compositor_chan.send(compositor).unwrap();
+    }
+
+    fn poll_events(&mut self, sender: IpcSender<bool>) {
+        let events = self.service.poll_events();
+        if events.len() > 0 {
+            let pipeline_ids: Vec<PipelineId> = self.contexts.iter().map(|c| *c).collect();
+            for event in events {
+                let event = WebVREventMsg::DisplayEvent(event);
+                self.constellation_chan.send(ConstellationMsg::WebVREvent(pipeline_ids.clone(), event)).unwrap();
+            }
+        }
+
+        // Stop polling events if the callers are not using VR
+        self.polling_events = self.contexts.len() > 0;
+        sender.send(self.polling_events).unwrap();
+    }
+
+    fn notify_event(&self, event: VRDisplayEvent) {
+        let pipeline_ids: Vec<PipelineId> = self.contexts.iter().map(|c| *c).collect();
+        let event = WebVREventMsg::DisplayEvent(event);
+        self.constellation_chan.send(ConstellationMsg::WebVREvent(pipeline_ids.clone(), event)).unwrap();
+    }
+
+    fn schedule_poll_events(&mut self) {
+        if self.service.is_initialized() && !self.polling_events {
+            self.polling_events = true;
+            let webvr_thread = self.sender.clone();
+            let (sender, receiver) = ipc::channel().unwrap();
+
+            thread::Builder::new().name("WebVRPollEvents".into()).spawn(move || {
+                loop {
+                    if webvr_thread.send(WebVRMsg::PollEvents(sender.clone())).is_err() {
+                        // WebVR Thread closed
+                        break;
+                    }
+                    if !receiver.recv().unwrap_or(false) {
+                        // WebVR Thread asked to unschedule this thread
+                        break;
+                    }
+                    thread::sleep(time::Duration::from_millis(500));
```

Comment:
This seems quite arbitrary. I understand why it's this way, but probably worth moving the `500` to a constant.